ion-hash/src/representation.rs

```diff
@@ -15,83 +17,87 @@ use ion_rs::{
+struct UpdateEscaping<'a, U: Update>(&'a mut U);
+
+impl<'a, U: Update> Update for UpdateEscaping<'a, U> {
+    fn update(&mut self, data: impl AsRef<[u8]>) {
+        for byte in data.as_ref() {
+            if let Markers::B | Markers::E | Markers::ESC = *byte {
+                self.0.update([0x0C]);
+            }
+            self.0.update([*byte]);
         }
     }
 }
 
-impl Representation for Option<&str> {
-    fn repr(&self) -> BinaryRepr {
-        match self {
-            Some(s) => Some(s.as_bytes().into()),
-            None => None,
-        }
+fn write_repr_integer<U: Update>(value: Option<&AnyInt>, hasher: &mut UpdateEscaping<'_, U>) {
+    match value {
+        Some(AnyInt::I64(v)) => match v {
+            0 => {}
+            _ => {
+                let magnitude = v.abs() as u64;
+                let encoded = binary::uint::encode_uint(magnitude);
+                hasher.update(encoded.as_bytes())
+            }
+        },
+        Some(AnyInt::BigInt(b)) => hasher.update(&b.magnitude().to_bytes_be()[..]),
+        None => {}
     }
 }
 
-impl Representation for Option<f64> {
-    /// Floats are encoded as big-endian octets of their IEEE-754 bit patterns,
-    /// except for special cases: +-zero, +-inf and nan.
-    fn repr(&self) -> BinaryRepr {
-        match self {
-            None => None,
-            Some(v) => {
-                // This matches positive and negative zero.
-                if *v == 0.0 {
-                    return if v.is_sign_positive() {
-                        None
-                    } else {
-                        Some(vec![0x80, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00])
-                    };
-                }
-                if v.is_infinite() {
-                    return if v.is_sign_positive() {
-                        Some(vec![0x7F, 0xF0, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00])
-                    } else {
-                        Some(vec![0xFF, 0xF0, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00])
-                    };
-                }
-
-                if v.is_nan() {
-                    return Some(vec![0x7F, 0xF8, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00]);
-                }
+/// Floats are encoded as big-endian octets of their IEEE-754 bit patterns,
+/// except for special cases: +-zero, +-inf and nan.
+fn write_repr_float<U: Update>(value: Option<f64>, hasher: &mut UpdateEscaping<'_, U>) {
+    match value {
+        None => {}
+        Some(v) => {
+            // This matches positive and negative zero.
+            if v == 0.0 {
+                return if !v.is_sign_positive() {
+                    hasher.update(&[0x80, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00])
+                };
+            }
+            if v.is_infinite() {
+                return if v.is_sign_positive() {
+                    hasher.update(&[0x7F, 0xF0, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00])
+                } else {
+                    hasher.update(&[0xFF, 0xF0, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00])
```

Comment:
*Minor*, it would be nice to have this in constants (magic numbers and all...)

Reply:
Done